server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java

```diff
@@ -0,0 +1,296 @@
+        }
+        final Map<String, Object> docMap = new ConcurrentHashMap<>(sourceMap);
+
+        // When a document completes processing, update the source with the inference
+        try (var docRef = new RefCountingRunnable(() -> {
+            if (docWriteRequest instanceof IndexRequest indexRequest) {
+                indexRequest.source(docMap);
+            } else if (docWriteRequest instanceof UpdateRequest updateRequest) {
+                if (updateRequest.docAsUpsert()) {
+                    updateRequest.upsertRequest().source(docMap);
+                } else {
+                    updateRequest.doc().source(docMap);
+                }
+            }
+            releaseOnFinish.close();
+        })) {
+
+            for (Map.Entry<String, Set<String>> fieldModelsEntrySet : fieldsForModels.entrySet()) {
+                String modelId = fieldModelsEntrySet.getKey();
+
+                @SuppressWarnings("unchecked")
+                Map<String, Object> rootInferenceFieldMap = (Map<String, Object>) docMap.computeIfAbsent(
+                    ROOT_INFERENCE_FIELD,
+                    k -> new HashMap<String, Object>()
+                );
+
+                List<String> inferenceFieldNames = getFieldNamesForInference(fieldModelsEntrySet, docMap);
+
+                if (inferenceFieldNames.isEmpty()) {
+                    continue;
+                }
+
+                InferenceProvider inferenceProvider = inferenceProvidersMap.get(modelId);
+                if (inferenceProvider == null) {
+                    failedItems.add(itemIndex);
+                    onBulkItemFailure.accept(
+                        bulkItemRequest,
+                        new IllegalArgumentException("No inference provider found for model ID " + modelId)
+                    );
+                    continue;
+                }
+                ActionListener<InferenceServiceResults> inferenceResultsListener = new ActionListener<>() {
+                    @Override
+                    public void onResponse(InferenceServiceResults results) {
+
+                        if (results == null) {
+                            throw new IllegalArgumentException(
+                                "No inference retrieved for model ID " + modelId + " in document " + docWriteRequest.id()
+                            );
+                        }
+
+                        int i = 0;
+                        for (InferenceResults inferenceResults : results.transformToLegacyFormat()) {
+                            String fieldName = inferenceFieldNames.get(i++);
+                            @SuppressWarnings("unchecked")
+                            List<Map<String, Object>> inferenceFieldResultList = (List<Map<String, Object>>) rootInferenceFieldMap
+                                .computeIfAbsent(fieldName, k -> new ArrayList<>());
+                            // Remove previous inference results if any
+                            inferenceFieldResultList.clear();
```

Comment:
If we always remove previous inference results, doesn't that mean we will re-run inference for every `semantic_text` field on a reindex?

Reply:
You're totally correct. We need additional logic to handle that.

This code is correct as when we receive back inference results, we want to remove the previous inference.

But, we should avoid to recalculate on reindex:
- We should always avoid calculating inference for an index action if there are already inference results.
- We should always recalculate inference for an update action for the included inference fields in the request

I'll work on that and add some more tests. Thanks!

Reply:
We could also break this out into a follow-up task if it makes the scope of this PR too big. It's already pretty chonky 😵‍💫